examples/oft/scripts/GasProfiler.s.sol

```diff
@@ -0,0 +1,243 @@
+    /// @param sender Sender address (OApp).
+    /// @param srcEid Source endpoint ID.
+    /// @param payload Message payload as a `bytes` array.
+    /// @param msgValue Ether value to send with the message (in wei).
+    /// @param numOfRuns Number of runs to execute.
+    function run_lzCompose(
+        string memory rpcUrl,
+        address endpointAddress,
+        uint32 srcEid,
+        address sender,
+        uint32 dstEid,
+        address receiver,
+        address composer,
+        bytes memory payload,
+        uint256 msgValue,
+        uint256 numOfRuns
+    ) external {
+        _initializeEndpoint(endpointAddress);
+        console.log("Starting gas profiling for lzCompose...");
+
+        TestParams memory params = _createTestParams(srcEid, sender, dstEid, receiver, payload, msgValue, numOfRuns);
+
+        _profileGasUsage(
+            rpcUrl,
+            params,
+            composer,
+            abi.encodeWithSelector(
+                ILayerZeroComposer(composer).lzCompose.selector,
+                params.receiver,
+                GUID.generate(
+                    1,
+                    params.srcEid,
+                    address(uint160(uint256(params.sender))),
+                    params.dstEid,
+                    bytes32(uint256(uint160(params.receiver)))
+                ),
+                params.payload,
+                address(this),
+                ""
+            )
+        );
+    }
+
+    /// @notice Initializes the endpoint contract.
+    /// @param endpointAddress Address of the LayerZero EndpointV2 contract.
+    function _initializeEndpoint(address endpointAddress) internal {
+        endpoint = ILayerZeroEndpointV2(endpointAddress);
+    }
+
+    /// @notice Creates a `TestParams` instance.
+    function _createTestParams(
+        uint32 srcEid,
+        address sender,
+        uint32 dstEid,
+        address receiver,
+        bytes memory payload,
+        uint256 msgValue,
+        uint256 numOfRuns
+    ) internal pure returns (TestParams memory) {
+        return
+            TestParams({
+                srcEid: srcEid,
+                sender: bytes32(uint256(uint160(sender))),
+                dstEid: dstEid,
+                receiver: receiver,
+                payload: payload,
+                msgValue: msgValue,
+                numOfRuns: numOfRuns
+            });
+    }
+
+    /// @notice Profiles gas usage of a function.
+    function _profileGasUsage(
+        string memory rpcUrl,
+        TestParams memory params,
+        address caller,
+        bytes memory callParams
+    ) internal {
+        uint256[] memory gasUsedArray = new uint256[](params.numOfRuns);
+        uint256 totalGasUsed = 0;
+        uint256 successfulRuns = 0;
+
+        for (uint256 i = 0; i < params.numOfRuns; i++) {
+            vm.createSelectFork(rpcUrl);
+            vm.deal(address(endpoint), 100 ether);
+            vm.startPrank(address(endpoint));
```

Comment:
@DanL0 this can't be outside of the loop because of the nature of Foundry forks.

The `vm.snapshot()` method does not revert state fully, so we need to take a fresh fork per run to ensure that state fully resets.